[UnifiedPDF] WebContent process jetsams when iterating through find results

hortont424 opened the pull request:
#### 0414cdeb4886bc95bbe545fe00bb1974dc9ed176
<pre>
[UnifiedPDF] WebContent process jetsams when iterating through find results
<a href="https://bugs.webkit.org/show_bug.cgi?id=295959">https://bugs.webkit.org/show_bug.cgi?id=295959</a>
<a href="https://rdar.apple.com/154292550">rdar://154292550</a>

Reviewed by Simon Fraser, Abrar Rahman Protyasha, and Aditya Keerthi.

There were a few flaws in 296356@main that resulted in it being less of a win
than expected:

Firstly, UnifiedPDFPlugin::rectsForTextMatch (called in a tight
loop by WebFoundTextRangeController::rectsForTextMatchesInRect) still
creates PDFKit objects for all text matches in order to compute their rects,
and only filters out the visible ones at the end of the process.

Secondly, `visibleRectsForFindMatchRects` deals with &quot;visibility&quot; based
on the current set of visible pages. But this doesn&apos;t make sense when being
asked to paint matches for a given tile - a tile could easily cover a page
that isn&apos;t currently visible, and nothing will cause it to later repaint
when the page becomes visible.

Fix these issues by computing the set of pages covered by the painting callback&apos;s
clip rect, instead of the &quot;currently visible&quot; ones, and filter the find matches
based on that page range *before* instantiating PDFKit objects (since that&apos;s
the expensive part).

This is a ~50x reduction in time spent under WebFoundTextRangeController::rectsForTextMatchesInRect
on a simple benchmark (iterating the first 100 matches in a long document with ~700 marked matches).

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::rectsForTextMatchesInRect const):
(WebKit::UnifiedPDFPlugin::rectsForTextMatchesInRect):
(WebKit::UnifiedPDFPlugin::rectsForTextMatch): Deleted.
Restructure rectsForTextMatch such that it takes a set of matches and returns all
the rects, instead of going one at a time. This is helpful because it means we can
compute the page range for the clip rect only once.

(WebKit::UnifiedPDFPlugin::visibleRectsForFindMatchRects const):
Filter by the passed-in clip rect instead of the unobscured rect.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.h:
(WebKit::PDFLayoutRow::intersectsPageRange const):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::rectsForTextMatchesInRect):
(WebKit::PluginView::rectsForTextMatch): Deleted.
* Source/WebKit/WebProcess/Plugins/PluginView.h:

* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::rectsForTextMatchesInRect):
Adapt the new `rectsForTextMatchesInRect` instead of `rectsForTextMatch`.

Canonical link: <a href="https://commits.webkit.org/297478@main">https://commits.webkit.org/297478@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f607314b8037dbe4e6bd8019a9c67c12365ccdea

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111872 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31535 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22012 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117890 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62100 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32216 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40116 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84979 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35667 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114819 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25730 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100668 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65419 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25051 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61731 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95109 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18874 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121164 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38899 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28921 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93849 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39280 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96925 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93672 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23918 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38862 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16641 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34910 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38797 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44292 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38434 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41759 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40149 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -438,11 +438,11 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
 
     Vector<WebFoundTextRange::PDFData> findTextMatches(const String&, WebCore::FindOptions) final;
-    Vector<WebCore::FloatRect> rectsForTextMatch(const WebFoundTextRange::PDFData&) final;
+    Vector<WebCore::FloatRect> rectsForTextMatchesInRect(const Vector<WebFoundTextRange::PDFData>&, WebCore::IntRect) final;
```
`const WebCore::IntRect&`

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -438,11 +438,11 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
     void scrollToRevealTextMatch(const WebFoundTextRange::PDFData&) final;
 
-    Vector<WebCore::FloatRect> visibleRectsForFindMatchRects(PDFPageCoverage) const;
+    Vector<WebCore::FloatRect> visibleRectsForFindMatchRects(PDFPageCoverage, WebCore::IntRect) const;
```
Ditto

smfr replied:
We should pass PDFPageCoverage by reference too.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3398,49 +3398,74 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
         return { };
 
+    IntRect clipRectInPluginSpace = convertFromRootViewToPlugin(clipRect);
+    auto clipRectInDocumentSpace = convertDown<FloatRect>(CoordinateSpace::Plugin, CoordinateSpace::PDFDocumentLayout, FloatRect { clipRectInPluginSpace });
```
Could you use `pageCoverageAndScalesForContentsRect` here instead?

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3398,49 +3398,74 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
+        if (!selection)
+            continue;
+
+        for (PDFPage *page in [selection pages]) {
```
RetainPtr

hortont424 replied:
does not compile, but also is needless because fast enumeration will throw an exception if the array is mutated (and the array holds a reference)

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3398,49 +3398,74 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
-    if (!selection)
+
+    RefPtr frame = m_frame.get();
+    if (!frame || !frame->page() || !frame->coreLocalFrame())
```
Do we need to check `!frame->page()`? I don't see it used anywhere?

hortont424 replied:
copy pasteo (though the original is wrong too)

hortont424 committed:
[UnifiedPDF] WebContent process jetsams when iterating through find results
https://bugs.webkit.org/show_bug.cgi?id=295959
rdar://154292550

Reviewed by Simon Fraser, Abrar Rahman Protyasha, and Aditya Keerthi.

There were a few flaws in 296356@main that resulted in it being less of a win
than expected:

Firstly, UnifiedPDFPlugin::rectsForTextMatch (called in a tight
loop by WebFoundTextRangeController::rectsForTextMatchesInRect) still
creates PDFKit objects for all text matches in order to compute their rects,
and only filters out the visible ones at the end of the process.

Secondly, `visibleRectsForFindMatchRects` deals with "visibility" based
on the current set of visible pages. But this doesn't make sense when being
asked to paint matches for a given tile - a tile could easily cover a page
that isn't currently visible, and nothing will cause it to later repaint
when the page becomes visible.

Fix these issues by computing the set of pages covered by the painting callback's
clip rect, instead of the "currently visible" ones, and filter the find matches
based on that page range *before* instantiating PDFKit objects (since that's
the expensive part).

This is a ~50x reduction in time spent under WebFoundTextRangeController::rectsForTextMatchesInRect
on a simple benchmark (iterating the first 100 matches in a long document with ~700 marked matches).

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::rectsForTextMatchesInRect const):
(WebKit::UnifiedPDFPlugin::rectsForTextMatchesInRect):
(WebKit::UnifiedPDFPlugin::rectsForTextMatch): Deleted.
Restructure rectsForTextMatch such that it takes a set of matches and returns all
the rects, instead of going one at a time. This is helpful because it means we can
compute the page range for the clip rect only once.

(WebKit::UnifiedPDFPlugin::visibleRectsForFindMatchRects const):
Filter by the passed-in clip rect instead of the unobscured rect.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.h:
(WebKit::PDFLayoutRow::intersectsPageRange const):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::rectsForTextMatchesInRect):
(WebKit::PluginView::rectsForTextMatch): Deleted.
* Source/WebKit/WebProcess/Plugins/PluginView.h:

* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::rectsForTextMatchesInRect):
Adapt the new `rectsForTextMatchesInRect` instead of `rectsForTextMatch`.

Canonical link: https://commits.webkit.org/297478@main
